src/section/shared_state.rs

```diff
@@ -314,6 +310,67 @@ impl SharedState {
         })
     }
 
+    /// Update our knowledge of their section and their knowledge of ours. Returns the events to
+    /// vote for (if any).
+    pub fn update_section_knowledge(
+        &mut self,
+        src: &SrcAuthority,
+        dst_key: Option<&bls::PublicKey>,
+        hash: &MessageHash,
+    ) -> Vec<AccumulatingEvent> {
+        let (&prefix, new_key) = if let Ok(pair) = src.as_section_prefix_and_key() {
+            pair
+        } else {
+            return vec![];
+        };
+
+        let is_neighbour = self.our_prefix().is_neighbour(&prefix);
+
+        let mut events = Vec::with_capacity(2);
```

Comment:
so, does it mean `AccumulatingEvent::TheirKeyInfo` and `AccumulatingEvent::TheirKnowledge` won't to be voted for at the same time?

Reply:
So the only possible combination of events that can be voted are these:
- `TheirKeyInfo` only
- `TheirKeyInfo` and `TheirKnowledge`
- `SendNeighbourInfo` only
- `SendNeighbourInfo` and `TheirKnowledge`
which is why it's enough to pre-allocate the vec to 2 elements.

Reply:
maybe better put the above as a comment in the code ?
it's really bit hard to figure this out by only reading the judgements  :)